src/lib/multicall/utils.ts

```diff
@@ -0,0 +1,131 @@
+import { Web3Provider } from "@ethersproject/providers";
+import { CHAIN_NAMES_MAP, getRpcUrl } from "config/chains";
+import { ContractCallContext, Multicall } from "ethereum-multicall";
+import { CallContext, ContractCallResults } from "ethereum-multicall/dist/esm/models";
+import { ethers } from "ethers";
+import { bigNumberify } from "lib/numbers";
+import { getFallbackProvider } from "lib/rpc";
+import { sleep } from "lib/sleep";
+import { MulticallRequestConfig, MulticallResult } from "./types";
+
+export const MAX_TIMEOUT = 2000;
+
+export async function executeMulticall(
+  chainId: number,
+  library: Web3Provider | undefined,
+  request: MulticallRequestConfig<any>
+) {
+  // Try to use rpc provider of the connected wallet
+  let provider = library ? library.getSigner().provider : undefined;
+
+  // Wait for initialization to chech the network
+  await provider?.ready;
+
+  // If the wallet is not connected or the network does not match the chainId of the request, create a new rpc provider
+  if (!provider || provider.network?.chainId !== chainId) {
+    const rpcUrl = getRpcUrl(chainId);
+
+    provider = new ethers.providers.StaticJsonRpcProvider(rpcUrl, { chainId, name: CHAIN_NAMES_MAP[chainId] });
+  }
+
+  const multicall = getMulticallLib(provider as ethers.providers.JsonRpcProvider);
```

Comment:
why it should be `JsonRpcProvider`? looks a bit strange. there is more generic `ethers.providers.Provider`, probably it may be used a  a type for `getMulticallLib` arg?